pkg/k8shandler/kibana/reconciler.go

```diff
@@ -0,0 +1,629 @@
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	kibanaServiceAccountName     = "kibana"
+	kibanaOAuthRedirectReference = "{\"kind\":\"OAuthRedirectReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"Route\",\"name\":\"kibana\"}}"
+)
+
+var (
+	kibanaServiceAccountAnnotations = map[string]string{
+		"serviceaccounts.openshift.io/oauth-redirectreference.first": kibanaOAuthRedirectReference,
+	}
+)
+
+func ReconcileKibana(requestCluster *kibana.Kibana, requestClient client.Client, proxyConfig *configv1.Proxy) error {
+	clusterKibanaRequest := ClusterKibanaRequest{
+		client:  requestClient,
+		cluster: requestCluster,
+	}
+
+	if clusterKibanaRequest.cluster == nil {
+		return nil
+	}
+
+	if err := clusterKibanaRequest.CreateOrUpdateServiceAccount(kibanaServiceAccountName, &kibanaServiceAccountAnnotations); err != nil {
+		return err
+	}
+
+	if err := clusterKibanaRequest.createOrUpdateKibanaService(); err != nil {
+		return err
+	}
+
+	if err := clusterKibanaRequest.createOrUpdateKibanaRoute(); err != nil {
+		return err
+	}
+
+	if err := clusterKibanaRequest.createOrUpdateKibanaConsoleExternalLogLink(); err != nil {
+		return err
+	}
+
+	if err := clusterKibanaRequest.createOrUpdateKibanaDeployment(proxyConfig); err != nil {
+		return err
+	}
+
+	kibanaStatus, err := clusterKibanaRequest.getKibanaStatus()
+	cluster := clusterKibanaRequest.cluster
+
+	if err != nil {
+		return fmt.Errorf("Failed to get Kibana status for %q: %v", cluster.Name, err)
+	}
+
+	printUpdateMessage := true
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry,
+		func() error {
+			if !compareKibanaStatus(kibanaStatus,
+				cluster.Status) {
+				if printUpdateMessage {
+					logrus.Infof("Updating status of Kibana")
+					printUpdateMessage = false
+				}
+				cluster.Status = kibanaStatus
+				return clusterKibanaRequest.UpdateStatus(cluster)
+			}
+			return nil
+		})
+	if retryErr != nil {
+		return fmt.Errorf("Failed to update Kibana status for %q: %v", cluster.Name, retryErr)
+	}
+	logrus.Infof("Kibana status successfully updated")
+
+	return nil
+}
+
+func ReconcileKibanaInstance(request reconcile.Request, rClient client.Client) error {
+	kibanaInstance := &kibana.Kibana{}
+	err := rClient.Get(context.TODO(), request.NamespacedName, kibanaInstance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	if kibanaInstance.Spec.ManagementState == kibana.ManagementStateUnmanaged {
+		return nil
+	}
+
+	if err := ReconcileKibana(kibanaInstance, rClient, getProxyConfig(rClient)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getProxyConfig(r client.Client) *configv1.Proxy {
+	proxyNamespacedName := types.NamespacedName{Name: constants.ProxyName}
+	proxyConfig := &configv1.Proxy{}
+	if err := r.Get(context.TODO(), proxyNamespacedName, proxyConfig); err != nil {
+		if !apierrors.IsNotFound(err) {
+			fmt.Printf("Encountered unexpected error getting %v. Error: %s\r\n", proxyNamespacedName, err.Error())
```

Comment:
if have have an error trying to get our proxy i feel we should actually be returning the error and requeueing the request

Reply:
Fixed